modules/apps/feature-flag/feature-flag-web/src/main/java/com/liferay/feature/flag/web/internal/configuration/DeprecationFeatureFlagsCompanyModelListener.java

```diff
@@ -0,0 +1,153 @@
+import com.liferay.portal.kernel.model.BaseModelListener;
+import com.liferay.portal.kernel.model.Company;
+import com.liferay.portal.kernel.model.CompanyConstants;
+import com.liferay.portal.kernel.model.ModelListener;
+import com.liferay.portal.kernel.service.CompanyLocalService;
+
+import java.io.IOException;
+
+import java.util.Arrays;
+import java.util.Dictionary;
+import java.util.Hashtable;
+import java.util.List;
+import java.util.Map;
+
+import org.osgi.framework.Constants;
+import org.osgi.framework.InvalidSyntaxException;
+import org.osgi.service.cm.Configuration;
+import org.osgi.service.cm.ConfigurationAdmin;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Drew Brokke
+ * @author Thiago Buarque
+ */
+@Component(
+	configurationPid = "com.liferay.feature.flag.web.internal.configuration.DeprecationFeatureFlags",
+	service = ModelListener.class
+)
+public class DeprecationFeatureFlagsCompanyModelListener
+	extends BaseModelListener<Company> {
+
+	@Override
+	public void onAfterCreate(Company company) throws ModelListenerException {
+		String filterString = StringBundler.concat(
+			"(", Constants.SERVICE_PID, "=",
+			DeprecationFeatureFlags.class.getName(), ")");
+
+		try {
+			Configuration[] configurations =
+				_configurationAdmin.listConfigurations(filterString);
+
+			if (configurations != null) {
+				Configuration configuration = configurations[0];
+
+				Dictionary<String, Object> properties =
+					configuration.getProperties();
+
+				_disableCompanyFeatureFlags(
+					company.getCompanyId(),
+					Arrays.asList(
+						(String[])properties.get("disabledFeatureFlags")));
+			}
+		}
+		catch (InvalidSyntaxException | IOException exception) {
+			throw new RuntimeException(exception);
+		}
+	}
+
+	@Activate
+	protected void activate(Map<String, Object> properties) {
+		try {
+			String filterString = StringBundler.concat(
+				"(", Constants.SERVICE_PID, "=",
+				DeprecationFeatureFlags.class.getName(), ")");
+
+			Configuration[] configurations =
+				_configurationAdmin.listConfigurations(filterString);
+
+			if ((configurations == null) && StartupHelperUtil.isDBNew()) {
+				_saveDeprecatedFeatureFlagsSnapshot();
+			}
+		}
+		catch (InvalidSyntaxException | IOException exception) {
+			throw new RuntimeException(exception);
+		}
+	}
+
+	private void _disableCompanyFeatureFlags(
+		long companyId, List<String> disabledFeatureFlags) {
+
+		CompanyFeatureFlags systemFeatureFlags =
+			_companyFeatureFlagsProvider.getOrCreateCompanyFeatureFlags(
+				CompanyConstants.SYSTEM);
+
+		List<FeatureFlag> deprecationFeatureFlags =
+			systemFeatureFlags.getFeatureFlags(
+				featureFlag ->
+					featureFlag.getFeatureFlagType(
+					).equals(
+						FeatureFlagType.DEPRECATION
+					) && disabledFeatureFlags.contains(featureFlag.getKey()));
```

Comment:
Brian would reject this because of method chaining. It might be better to just expand this into a method body with `{ }` curly braces.